File: 72-edit-distance/edit-distance.java
class Solution {
    private int solve(int i, int j, String s, String t, int[][] dp){
        if(j==t.length()) return s.length()-i;
        if(i==s.length()) return t.length()-j;

        if(dp[i][j] != -1) return dp[i][j];
        int insert =0, delete = 0,replace = 0;
        if(s.charAt(i)==t.charAt(j)){
            return solve(i+1,j+1,s,t,dp);
        }else{
            insert = solve(i,j+1,s,t,dp);
            delete = solve(i+1,j,s,t,dp);
            replace = solve(i+1,j+1,s,t,dp);
            return dp[i][j] = 1 + Math.min(insert, Math.min(delete,replace));
        }
    }
    public int minDistance(String word1, String word2) {
        int[][] dp = new int[word1.length()][word2.length()];
        for(int i=0;i<word1.length();i++){
            Arrays.fill(dp[i],-1);
        }
        return solve(0,0,word1,word2,dp);
    }
}

File: 72-edit-distance/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n, m = len(word1), len(word2)

        dp = {}

        def solve(i, j):
            if j < 0:
                return i + 1

            if i < 0:
                return j + 1

            if (i, j) in dp:
                return dp[(i, j)]

            if word1[i] == word2[j]:
                dp[(i, j)] = solve(i - 1, j - 1)

            else:
                dp[(i, j)] = 1 + min(
                    solve(i - 1, j),       # delete
                    solve(i - 1, j - 1),   # replace
                    solve(i, j - 1)       # insert
                )

            return dp[(i, j)]

        return solve(n - 1, m - 1)